src/base/bittorrent/session.cpp

```diff
@@ -1350,6 +1358,57 @@ bool Session::addTorrent_impl(AddTorrentData addData, const MagnetUri &magnetUri
     return true;
 }
 
+bool Session::findIncompleteFiles(TorrentInfo &torrentInfo, QString &savePath) const
+{
+    auto findInDir = [](const QString &dirPath, TorrentInfo &torrentInfo) -> bool
+    {
+        bool found = false;
+        if (torrentInfo.filesCount() == 1) {
+            const QString filePath = dirPath + torrentInfo.filePath(0);
+            if (QFile(filePath).exists()) {
+                found = true;
+            }
+            else if (QFile(filePath + QB_EXT).exists()) {
+                found = true;
+                torrentInfo.renameFile(0, torrentInfo.filePath(0) + QB_EXT);
+            }
+        }
+        else {
+            QSet<QString> allFiles;
+            int dirPathSize = dirPath.size();
+            foreach (const QString &file, findAllFiles(dirPath + torrentInfo.name()))
+                allFiles << file.mid(dirPathSize);
+            for (int i = 0; i < torrentInfo.filesCount(); ++i) {
+                QString filePath = torrentInfo.filePath(i);
+                auto it = allFiles.find(filePath);
+                if (it != allFiles.end()) {
+                    found = true;
+                    allFiles.erase(it); // optimize for further search speedup
```

Comment:
Maybe this erasing is overkill as there are really few chances to have similar hashes...